refactor : implement chart title with options (#3)

Rengod95 opened the pull request:
**작업 이슈**
- https://github.com/ChartStudio/Dynamic_Chart/issues/3
- Tooltip Commit 기준 코드 마이그레이션

**작업 상세 설명**
- title parameter의 다형성 제공
  - title 의 undefiend, string, TitleConfig 3가지 형태에 대한 constructor 지원
  - test.html 에서 다양한 형태의 title 입력받기 가능

- TitleConfig의 Options 객체 수정
  - nullable -> unable null
  - DEFAULT Options 객체 추가

- 오타 및 수정 사항
  - TaskEventQueue 의 regist -> register 로 수정
  - TaskEventQueue 의 buildTaskEvent -> createTaskEvent 로 수정
    - build 단어가 추상적이라 문맥상 더 바른 Create로 수정
    - TaskRunner 의 buildTaskRunner 도 동치 
  - 메소드 별 간단한 주석 일부 추가

**참조 문서**
- N/A

Rengod95 committed:
refactor : implement chart title with options (#3)

- Tooltip Commit 기준 코드 마이그레이션

- title parameter의 다형성 제공
  - title 의 undefiend, string, TitleConfig 3가지 형태에 대한 constructor 지원
  - test.html 에서 다양한 형태의 title 입력받기 가능

- TitleConfig의 Options 객체 수정
  - nullable -> unable null
  - DEFAULT Options 객체 추가

- 오타 수정 사항
  - TaskEventQueue 의 regist -> register 로 수정
  - TaskEventQueue 의 buildTaskEvent -> createTaskEvent 로 수정
    - build 단어가 추상적이라 문맥상 더 바른 Create로 수정
    - TaskRunner 의 buildTaskRunner 도 동치
  -
  - 메소드 별 간단한 주석 일부 추가

Rengod95 committed:
refactor : implement chart title with options (#3)

- type.chart.title.ts 부분 수정

MK-Lee13 commented on src/task/task.graph.controller.ts:
```diff
@@ -19,87 +19,87 @@ class GraphTaskController {
 
   init() {
-    this.taskEventQueue.regist(
+    this.taskEventQueue.register(
```
개인적인 생각으로 regist 가 더 맞는 단어라고 생각합니다. register 의 경우 등록자 혹은 컴퓨터 부품을 의미하는 반면
regist 의 경우 등록하다라는 행위를 표현합니다. 특히 함수의 경우 내가 어떠한 행위를 하는지를 설명해야하는 데 이 경우 register 는 맞지 않는 느낌이 듭니다.

MK-Lee13 commented on src/core/task/task.runner.ts:
```diff
@@ -9,7 +9,7 @@ class TaskRunner {
   }
 
-  static buildTaskRunner(parameters: any[] | null, task: Function) {
+  static createTaskRunner(parameters: any[] | null, task: Function) {
```
이렇게 객체를 생성해주는 함수를 `정적 팩토리 메서드` 라고 합니다. 네이밍이 빌드 보다는 크리에이트가 맞는 것 같네요. 수고하셨습니다.

MK-Lee13 commented on src/core/core.config.ts:
```diff
@@ -68,6 +74,7 @@ class BaseConfig {
   verticalConfig: VerticalStyleConfig;
   backgroundConfig: BackgroundStyleConfig;
   tooltipConfig: TooltipStyleConfig;
+  chartTitleConfig:ChartTitleConfig;
```
한칸만 뛰어 주세요.

MK-Lee13 commented on src/core/task/task.event.queue.ts:
```diff
@@ -24,8 +24,9 @@ class TaskEventQueue {
-  regist(type: string, parameters: any[] | null, task: Function) {
-    this.queue.push(this.buildTaskEvent(type, parameters, task))
+  register(type: string, parameters: any[] | null, task: Function) {
+    //콜백과 콜백의 prams를 받아 TaskEvent를 생성하고 실행
```
regist 가 맞는 표현으로 보입니다. 결국 이벤트 큐에 등록을 하는 작업을 의미하는 함수라서 등록자 보다는 등록하다가 맞다고 생각합니다.

MK-Lee13 commented on src/core/task/task.event.queue.ts:
```diff
@@ -68,6 +72,7 @@ class TaskEventQueue {
       return true;
     }
 
+    //테스크가 생성된 시점과 excute 되는 시점의 차이가 스킵 단위보다 크면 정상적인 테스크
```
이벤트 큐에 작성하신 주석 제거 부탁드립니다.

MK-Lee13 commented on src/view/view.graph.ts:
```diff
@@ -57,7 +62,7 @@ class GraphView {
     let image = this.styler.getBackgroundImage()
 
-    this.context?.drawImage(image, 0, 0, pixel.x, pixel.y);
+    this.context?.drawImage(image, 0, 0, pixel.x+200, pixel.y+200);
```
200을 왜 붙이신 건지 알 수 있을까요? 또한 만약 정말 필요한 작업이라면 따로 함수를 작성하셔야 합니다.

Rengod95 replied:
커밋이 꼬여 이전에 변경했던 사항이 추가된 것으로 확인됩니다. 바로 수정하도록 하겠습니다!

MK-Lee13 commented on src/core/core.config.ts:
```diff
@@ -91,6 +98,7 @@ class BaseConfig {
     this.type = options.type;
     this.width = options.width ?? DEFAULT_WIDTH
     this.height = options.height ?? DEFAULT_HEIGHT
+    this.height += DEFAULT_EXTRA_HEIGHT_FOR_TITLE
```
해당 케이스에서 가장 궁금한 것은 저 경우 만약 제목을 붙이고 싶지 않은 경우에도 무조건 높이를 증가시키지 않나요?
혹은 아직 제목 제거 부분은 구현을 아직 안한건가요??

MK-Lee13 replied:
해당 부분이 사용자가 입력한 데이터 값의 변화를 주는 부분이군요. 해당 부분은 삭제하셔야 합니다.
사용자가 입력한 width, height 에 맞게 캔버스의 변화가 생겨야하며, 이외의 부분은 Y Margin 을 통해 해결하셔야 합니다.
결국 이 경우에는 내가 height 를 40 을 줘도 결국 결과물은 저희가 임의로 개산한 60 이 나오기 때문에 해당 부분 변경 부탁드립니다.

Rengod95 replied:
-해당 케이스에서 가장 궁금한 것은 저 경우 만약 제목을 붙이고 싶지 않은 경우에도 무조건 높이를 증가시키지 않나요?
혹은 아직 제목 제거 부분은 구현을 아직 안한건가요??

A. 작업이 미숙했습니다. TitleConfig 생성 코드를 전반적으로 변경하면서 해당 부분도 수정하여, title이 존재하지 않을 땐, 사이즈를 일절 건드리지 않도록 수정하였습니다.

-사용자가 입력한 width, height 에 맞게 캔버스의 변화가 생겨야하며, 이외의 부분은 Y Margin 을 통해 해결하셔야 합니다.
A. 
동의합니다. 사용자가 입력한 width, height는 결국 '차트'의 크기라고 생각했습니다.'

그러나 현재 제가 파악한 바로는 사용자에게 입력받은 width, height가 '캔버스'의 크기 및 '차트'의 크기 모두에 사용되고 있습니다.
즉 width:600, height: 400 을 사용자에게 입력받는다고 가정하면, 브라우저 내에서 canvas element가 차지하는 영역과, 차트의 크기가 동일합니다.

Position 객체에서 차트의 Height를 결정하는 로직을 완벽하게 파악하진 않았으나, 
this.height - this.baseYAxisMargin - this.baseYAxisPadding 라고 알고 있습니다.
여기서 this.height는 사용자에게 입력받은 height 값을 그대로 사용한다고 파악됩니다.

즉, 말씀하신대로 Y margin 혹은 Y padding 을 사용한다고 해도,  결국 '차트'의 height에 변화가 발생합니다.

결국 차트의 크기에 영향을 주지 않도록 하기 위해서는, canvas element의 크기와, 차트의 크기를 분리하는 것과 같은 수정이 필요하다고 생각했습니다.

MK-Lee13 replied:
차트가 변동되는 것은 의미가 없습니다. canvas가 고정이되어야 합니다.
해당 라이브러리를 사용하는 실제 사용자의 경우 개발자 입니다. 이때 만약 내가 600 * 400 으로 선언하고 결과값을 보니 
실제 크기가 600 * 460 이 되면 안됩니다. 그렇기 차트의 height 는 중요하지 않습니다.

우리가 실제 차트 height 를 보고 디자인하지 못합니다. 나는 차트 크기만 height 몇으로 해야지 하는 경우는 없고 차트가 들어갈 공간에 크기를 보기 때문에 이 부분에서 height 의 값 결과를 임의로 변경하는 작업은 제거 되어야 합니다.

MK-Lee13 reviewed (commented):
수고하셨습니다~~👍 리뷰 확인해주시고 수정해주시면 approve 승인 후 머지하시면 되겠습니다.
그전에 PR 3번 close 해주세요. 해당 PR 은 현재 머징될 일이 없기 때문에 close 하셔야 합니다.

Rengod95 committed:
fix : implement chart title with options (#3)

- createTitleConfig 메소드 추가
- ChartTitleConfig 의 생성자 간소화
- BaseConfig 의 Title 유무에 따른 height 증가 로직 추가
- ChartTitle Interface 수정
  - 기존 type->titleType 으로 변경
  - 고유 type 을 "ChartTitle"로 수정

Rengod95 committed:
fix : implement chart title with options (#3)

- createTitleConfig 메소드 추가
- ChartTitleConfig 의 생성자 간소화
- BaseConfig 의 Title 유무에 따른 height 증가 로직 추가
- ChartTitle Interface 수정
  - 기존 type->titleType 으로 변경
  - 고유 type 을 "ChartTitle"로 수정

MK-Lee13 commented on src/core/core.config.ts:
```diff
@@ -74,6 +74,7 @@ class BaseConfig {
   backgroundConfig: BackgroundStyleConfig;
   tooltipConfig: TooltipStyleConfig;
+
   chartTitleConfig:ChartTitleConfig;
```
```
chartTitleConfig:ChartTitleConfig; // -> X
chartTitleConfig: ChartTitleConfig; // -> O
```
이걸 의미한거였습니다.

Rengod95 committed:
fix : implement chart title with options (#3)

- createTitleConfig 메소드 추가
- ChartTitleConfig 의 생성자 간소화
- BaseConfig 의 Title 유무에 따른 height 증가 로직 추가
- ChartTitle Interface 수정
  - 기존 type->titleType 으로 변경
  - 고유 type 을 "ChartTitle"로 수정

MK-Lee13 reviewed (approved):
작업하시느라 고생하셨습니다~~~~👍
